shiny/ui/_theme.py

```diff
@@ -0,0 +1,364 @@
+        # so that we can re-use the already compiled CSS file.
+        self._css_temp_srcdir: Optional[str] = None
+
+    @classmethod
+    def available_presets(cls) -> tuple[ShinyThemePreset, ...]:
+        """
+        Get a list of available theme presets.
+        """
+        return shiny_theme_presets
+
+    @property
+    def preset(self) -> ShinyThemePreset:
+        return self._preset
+
+    @preset.setter
+    def preset(self, value: ShinyThemePreset) -> None:
+        check_is_valid_preset(value)
+        self._preset = value
+
+        has_customizations = (
+            len(self._functions) > 0
+            or len(self._defaults) > 0
+            or len(self._mixins) > 0
+            or len(self._rules) > 0
+        )
+
+        if has_customizations:
+            self._css = ""
+        else:
+            self._css = "precompiled" if value in shiny_theme_presets_bundled else ""
+
+        self._css_temp_srcdir = None
+
+    def add_functions(self: T, *args: str) -> T:
+        """
+        Add custom Sass functions to the theme.
+
+        Sass code added via this method will be placed **after** the function
+        declarations from the theme preset, allowing you to override or extend the
+        default functions.
+
+        Parameters
+        ----------
+        *args
+            The Sass functions to add as a single or multiple strings.
+        """
+        self._css = ""
+        self._functions.extend(dedent_array(args))
+        return self
+
+    def add_defaults(
+        self: T, *args: str, **kwargs: str | float | int | bool | None
+    ) -> T:
+        """
+        Add custom default values to the theme.
+
+        Sass code added via this method will be placed **before** the default values of
+        the theme preset, allowing you to override or extend the default values.
+
+        Parameters
+        ----------
+        *args
+            Sass code, as a single or multiple strings, containing default value
+            declarations to add.
+        **kwargs
+            Keyword arguments containing default value declarations to add. The keys
+            should be Sass variable names using underscore casing that will be
+            transformed automatically to kebab-case. For example,
+            `.add_defaults(primary_color="#ff0000")` is equivalent to
+            `.add_defaults("$primary-color: #ff0000;")`.
+        """
+        if len(args) > 0 and len(kwargs) > 0:
+            # Python forces positional arguments to come _before_ kwargs, but default
+            # argument order might matter. To be safe, we force users to pick one order.
+            raise ValueError("Cannot provide both positional and keyword arguments")
+        elif len(args) == 0 and len(kwargs) == 0:
+            return self
+
+        defaults: list[str] = list(args)
+
+        if len(kwargs) > 0:
+            for key, value in kwargs.items():
+                key.replace("_", "-")
```

Comment:
I think we'll want a way to opt-out of this since `_` is valid in Sass variable names

Reply:
The way out is to use positional args, e.g. `.add_defaults("$my_default: foo")`